codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/XmlParserGenerator.kt

```diff
@@ -152,10 +184,440 @@ open class XmlParserGenerator(
+                } else {
+                    deserializeMap(ctx, serdeCtx, member, writer)
+                }
+            }
+            ShapeType.STRUCTURE, ShapeType.UNION -> {
+                val deserializeFn = documentDeserializer(ctx, target)
+                writer.write("#T(${serdeCtx.tagReader})", deserializeFn)
+            }
+            else -> deserializePrimitiveMember(
+                ctx,
+                member,
+                writer.format("${serdeCtx.tagReader}.#T()", SerdeXml.tryData),
+                textExprIsResult = true,
+                writer,
+            )
+        }
+    }
+
+    // TODO - this could probably be moved to SerdeExt and commonized
+
+    private fun Shape.shapeDeserializerDefinitionFile(
+        ctx: ProtocolGenerator.GenerationContext,
+    ): String {
+        val target = targetOrSelf(ctx.model)
+        val shapeName = StringUtils.capitalize(target.id.getName(ctx.service))
+        return "${shapeName}ShapeDeserializer.kt"
+    }
+    private fun Shape.shapeDeserializer(
+        ctx: ProtocolGenerator.GenerationContext,
+        block: (fnName: String, writer: KotlinWriter) -> Unit,
+    ): Symbol {
+        val target = targetOrSelf(ctx.model)
+        val shapeName = StringUtils.capitalize(target.id.getName(ctx.service))
+        val symbol = ctx.symbolProvider.toSymbol(this)
+
+        val fnName = "deserialize${shapeName}Shape"
+        return buildSymbol {
+            name = fnName
+            namespace = ctx.settings.pkg.serde
+            definitionFile = shapeDeserializerDefinitionFile(ctx)
+            reference(symbol, SymbolReference.ContextOption.DECLARE)
+            renderBy = {
+                block(fnName, it)
+            }
+        }
+    }
+
+    private fun deserializeShape(
+        ctx: ProtocolGenerator.GenerationContext,
+        shape: Shape,
+        block: KotlinWriter.() -> Unit,
+    ): Symbol {
+        val symbol = ctx.symbolProvider.toSymbol(shape)
+        val deserializeFn = shape.shapeDeserializer(ctx) { fnName, writer ->
+            writer.withBlock(
+                "internal fun #L(reader: #T): #T {",
+                "}",
+                fnName,
+                SerdeXml.XmlTagReader,
+                symbol,
+            ) {
+                block(this)
+            }
+        }
+        return deserializeFn
+    }
+
+    private fun deserializeList(
+        ctx: ProtocolGenerator.GenerationContext,
+        serdeCtx: SerdeCtx,
+        member: MemberShape,
+        writer: KotlinWriter,
+    ) {
+        val target = ctx.model.expectShape<CollectionShape>(member.target)
+        val targetMember = target.member
+        val isSparse = target.hasTrait<SparseTrait>()
+        val deserializeFn = deserializeShape(ctx, target) {
+            write("val result = mutableListOf<#T#L>()", ctx.symbolProvider.toSymbol(targetMember), nullabilitySuffix(isSparse))
+            deserializeLoop(SerdeCtx(tagReader = "reader")) { innerCtx ->
```

Comment:
Right now no but if we add additional state at some point then it _might_ matter. The reason we construct a new one in a few spots (like list/map deserialization) is because those are actually in entire new functions like this:


```kotlin
internal fun deserializeSomeListMember(reader: TagReader): List<Foo> {
    ...
}
```

So the serde context is reset with a new `tagReader` named `reader`. Whereas in other parts of the codegen we are still inside the same scope as the `SerdeCtx` we were given.


```kotlin
val root = xmlTagReader(payload)

loop@while (true) {
     val curr = root.nextTag() ?: break@loop    // this SerdeCtx is inside the same scope as the original SerdeCtx(root)
     when (curr.tagName) { 
          "MyListMember" -> builder.foos = deserializeSomeListMember(curr) // we enter a new serde ctx
     }
}

```

In practice there isn't much difference at the moment but we are constructing new ones whenever we are inside a new deserializer function basically.